docs/adr/0002-delivery-channel-creation-and-integration.md

```diff
@@ -0,0 +1,81 @@
+# New delivery channel creation and integration
+
+## Status
+
+Proposed
+
+## Context
+
+Delivering notifications through different communication types (email, sms, web push notifications etc) is necessary in an extendable notifications service. Users should be able to choose which delivery channels (twilio, sendgrid, airship etc) they want to use for the associated communication type(s) as these said platforms could have different features or costs and if combined could provide less setup configuration/cost, which leaves the configuration totally up to the end user. 
+
+## Decision
+```
+                 +------------+
+                 |            |
+                 | Management |
+                 | API        |
+                 |            |
+                 |            |
+                 ++--+------+-+
+                  ^  ^      ^
+                  |  |      |
+         +--------+  |      +--------------+
+         |           |                     |
+         v           v                     v
++--------+-+      +--+-------+           +-+--------+
+|          |      |          |           |          |
+| Delivery |      | Delivery |           | Delivery |
+| Channel  |      | Channel  | +-+  +-+  | Channel  |
+| 1        |      | 2        |           | N        |
+|          |      |          |           |          |
+|          |      |          |           |          |
++----------+      +----------+           +----------+
+```
+
+The management api should have knowledge of what delivery channels are available and where certain notification types should be sent. As we send notifications through on a per user per tenant basis, user settings for which communication type(s) a user has allowed is read and a notification is sent to the available delivery channel(s). 
+
+### User notifications and user settings
+
+A user should be able to choose what kind of communication types they receive based on the notification type. There are two ways of approaching this, either:
+1. Send different types of notifications through all channels that are provided by the consumer of this service, and leave the communication type and user configuration for the consumer of the Marain.UserNotifications service. (ie, the consumer of the Marain.UserNotifications service should specify which channels they want to send notifications via an api call when a notification is created)
+2. Store user preferences somewhere that is accessible to the management api, then when a new type of notifications comes through, check what type of communications types  the user has agreed to use and then send a message through the delivery channel .
```

Comment:
My initial thought was how would we handle say a transactional email vs an email of importance. 

I agree with you, adding a severity to be used in conjunction with the notification type will allow a user of `Marain.UserNotificaitons` to specify that they want to receive a certain type of severity through a certain channel while a different one through a different channel. A default severity would be set is not provided.

I'm not very sure what you mean by the statement:
>Yet other apps might want something more complex, e.g., a hierarchical categorisation. And some might want more than 2 dimensions.
Do you mean something along the lines of having different layers of severity? Do you mean having a something that could be severe that only needs to be communicated to a few individuals based on a certain dimension? That sounds like that should be worked out before it's even sent to the `Marain.UserNotifications`.

Reply:
By "hierarchichal categorisation" I meant that you might have some top level categories, e.g., "Build" "Release" and "Repositories". And then within that you might have subcategories, e.g, "Build/Execution", "Build/Configuration Changes" etc. And these might be further subdivded, e.g. "Build/Execution/Start", "Build/Execution/Progress", "Build/Execution/Complete".

As for dimensions, in addition to being able to filter by severity, you might want to filter based on some other property. E.g., I might want to see all build notifications for my own team, but I might want to see only build failures for other teams. So some "team" property is now an additional "dimension" for filtering in the sense I was using it above.

Reply:
I don't think we need to go that far right now.

I do agree that it will need to be done (hierarchical categorisation), but I think that should be either split into a different ADR (so it can be more well defined) or let the application developer that integrates this platform into their system handle that logic.

Do we want to handle everything or just handle what notification type/communication type/severity that users want to use. It might be get quite complicated for someone who wants to integrate Marain.UserNotifications into their system in the future if we do head down this route.

Reply:
I agree that we don't want to go too far, but the point I was trying to make was that by selecting 2 you are effectively committing us to baking some kind of model of what a notification type is directly into the system. I am not convinced we understand the problem well enough to get this right, which is why I would favour the idea that the application that consumes this service is responsible for deciding which messages get sent to who and via which channels.

But if we are going to make `Marain.UserNotifications` itself take the decisions about who gets what based on configuration settings, I think we need a more concrete example. I'd want it to be specific in two ways. First, I'd want application-specific requirements, ideally a description of the cases you want to support in your application. Second, I'd want to see examples of what we thing the configuration for that in `Marain.UserNotifications` would actually look like.

Reply:
The service that consumes `Marian.UserNotifications` already decides which messages are sent to who. The channels which the notification type will be sent through is the topic we're discussing. 

A concrete example of a notification being sent through the `Marain.UserNotifications` system would be as follows.
1. A CreateNotificationRequest object is sent to the `Marain.UserNotifications.Management.Host` by an external consuming system.
2. A User preference (based on the userId attached to the CreateNotificationRequest object) is read  and included in that user preference are the different communication types that the user has allowed for the current notification type (pre-configured by a user). The notification type is just a string, so severity can be included in the notification type if that is needed right now.
3. Based on the communication types, different objects will be sent to the configured delivery channels (this would include templating everything etc).